### DISCUSSION TITLE
JDKs older than JDK 21 are not supported anymore effective Bazel 7.1 and onwards

### DISCUSSION THREAD

**Discussion by lberki** [bazelbuild/bazel#21446] [Announcement]
In the future, Bazel will need a JVM 21 or later to run.

Given that Bazel is ordinarily packaged with a JVM it runs under by default, this will not be visible change for the majority of our users. Only people who use `bazel_nojdk` (which requires one to supply their own JVM) and people who build their own Bazel will be affected.

The policy will take effect in Bazel 7.1 and Bazel at HEAD as of now. Initially, the code will still run with older JDKs, but we'll probably introduce a dependency on JVM 21 pretty soon at which point this will not be the case anymore.

For any questions of concerns, follow up on this thread; if you have a use case that is broken by this change, we'll try to find a solution.

**Comment by siepkes** [comment]
> Given that Bazel is ordinarily packaged with a JVM it runs under by default, this will not be visible change for the majority of our users. Only people who use bazel_nojdk (which requires one to supply their own JVM) and people who build their own Bazel will be affected.

That's basically everyone who is on platforms such as FreeBSD.

**Reply by meisterT** [reply]
Looking at https://www.freshports.org/java/openjdk21/, it seems to me that JDK21 is available on FreeBSD. Is that assumption incorrect?

**Reply by siepkes** [reply]
You are correct. I see now that since 3 months there is a FreeBSD OpenJDK 21 port. So my question to you, did you folks check that and factor that in to your decision? And what about other platforms?

**Reply by meisterT** [reply]
I indeed checked FreeBSD when Lukacs brought up the discussion. I believe that most users are using Bazel with the embedded JDK anyway, but I could be mistaken. Any other platform you are worried about?

**Reply by siepkes** [reply]
No particular OS'es spring to mind. I initially thought even FreeBSD wasn't there yet. Though I do suspect for most niche platforms (Haiku, etc.) OpenJDK 21 might be a pretty high bar to reach. Since Loom is in 21.

**Reply by lberki** [reply]
FWIW, the reason why we decided to require JDK 21 is exactly Loom. It's very useful. Is there a reason to think that Loom is a big hindrance to supporting JDK 21 on niche OSes?

JDK 21 was released to GA in 2023 September (so five months ago) so I'd think that it should be enough time to port it to these niche OSes.

**Reply by siepkes** [reply]
> Is there a reason to think that Loom is a big hindrance to supporting JDK 21 on niche OSes?

Loom depends on (native) code in the JVM. Native code which previously didn't exist in the JVM. Therefor I suspect 21 has more native code to port then most previous releases. I have no idea how big of  a problem this would actually be for these niche OS'es to port. What I've seen most of this code revolves around non-blocking stuff so for example in Linux it plugs into epoll.

Initially I thought even FreeBSD hadn't updated their port yet. So the issue is definitely smaller then I initially thought.

**Comment by zcdlab** [comment]
> The policy will take effect in Bazel 7.1 and Bazel at HEAD as of now. Initially, the code will still run with older JDKs, but we'll probably introduce a dependency on JVM 21 pretty soon at which point this will not be the case anymore.

I'm upgrading Bazel from 7.1.1 to 7.4.1 while maintaining the old JDK 11 in this repository. When I build the project, I encounter the following error. How can I resolve this issue?

```
Error: LinkageError occurred while loading main class com.google.devtools.build.lib.bazel.Bazel
        java.lang.UnsupportedClassVersionError: com/google/devtools/build/lib/bazel/Bazel has been compiled by a more recent version of the Java Runtime (class file version 65.0), this version of the Java Runtime only recognizes class file versions up to 55.0
 ```
